modules/caddyhttp/reverseproxy/upstreams.go

```diff
@@ -0,0 +1,243 @@
+package reverseproxy
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/caddyserver/caddy/v2"
+)
+
+func init() {
+	caddy.RegisterModule(SRVUpstreams{})
+	caddy.RegisterModule(AUpstreams{})
+}
+
+// SRVUpstreams provides upstreams from SRV lookups.
+// The lookup DNS name can be configured either by
+// its individual parts (that is, specifying the
+// service, protocol, and name separately) to form
+// the standard "_service._proto.name" domain, or
+// the domain can be specified directly in name by
+// leaving service and proto empty. See RFC 2782.
+//
+// Lookups are cached and refreshed at the configured
+// refresh interval.
+//
+// Returned upstreams are sorted by priority and weight.
+type SRVUpstreams struct {
+	// The interval at which to refresh the SRV lookup.
+	// Results are cached between lookups. Default: 1m
+	Refresh time.Duration `json:"refresh,omitempty"`
+
+	// The service label.
+	Service string `json:"service,omitempty"`
+
+	// The protocol label; either tcp or udp.
+	Proto string `json:"proto,omitempty"`
+
+	// The name label; or, if service and proto are
+	// empty, the entire domain name to look up.
+	Name string `json:"name,omitempty"`
+}
+
+// CaddyModule returns the Caddy module information.
+func (SRVUpstreams) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "http.reverse_proxy.upstreams.srv",
+		New: func() caddy.Module { return new(SRVUpstreams) },
+	}
+}
+
+// String returns the RFC 2782 representation of the SRV domain.
+func (su SRVUpstreams) String() string {
+	return fmt.Sprintf("_%s._%s.%s", su.Service, su.Proto, su.Name)
+}
+
+func (su *SRVUpstreams) Provision(_ caddy.Context) error {
+	if su.Proto != "tcp" && su.Proto != "udp" {
+		return fmt.Errorf("invalid proto '%s'", su.Proto)
+	}
+	if su.Refresh == 0 {
+		su.Refresh = time.Minute
+	}
+	return nil
+}
+
+func (su SRVUpstreams) GetUpstreams(r *http.Request) ([]*Upstream, error) {
+	suStr := su.String()
+
+	// first, use a cheap read-lock to return a cached result quickly
+	srvsMu.RLock()
+	cached := srvs[suStr]
+	srvsMu.RUnlock()
+	if cached.fresh() {
+		return cached.upstreams, nil
+	}
+
+	// otherwise, obtain a write-lock to update the cached value
+	srvsMu.Lock()
+	defer srvsMu.Unlock()
+
+	// check to see if it's still stale, since we're now in a different
+	// lock from when we first checked freshness; another goroutine might
+	// have refreshed it in the meantime before we re-obtained our lock
+	cached = srvs[suStr]
+	if cached.fresh() {
+		return cached.upstreams, nil
+	}
+
+	// prepare parameters and perform the SRV lookup
+	repl := r.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
+	service := repl.ReplaceAll(su.Service, "")
+	proto := repl.ReplaceAll(su.Proto, "")
+	name := repl.ReplaceAll(su.Name, "")
+
+	_, records, err := net.DefaultResolver.LookupSRV(r.Context(), service, proto, name)
```

Comment:
I think if someone wants to use the SRV results by priority, they'd have to use the `first` selection policy, and also make sure to have passive health checks enabled so that Caddy actually marks the top priority backends as unhealthy so it actually tries lower priorities.

Because otherwise, the default of `random` would just do whatever, which doesn't respect the SRV response.

Also, if we're caching the result, we need to keep in mind that we're also caching the result of the RNG that happened to randomize between the weights. So only one backend out of a given priority will get routed to while the result is cached. That seems like it might be a problem.

It is kinda nice that the stdlib does this for us, but I think their assumption is that these results don't get cached and instead this gets queried every time.

Should we have two implementations of `SRVUpstreams`? One which is cached, and _doesn't_ properly support weights, and another that's not cached, and _does_ properly support weights.

Correct me if I'm wrong here, I might've made an incorrect assumption.

Reply:
Yes, I think if you want SRV load balancing then using `first` policy with passive health checks is the right idea, and is the same way this would be done with static upstreams too.

You're right that the caching of results also caches the randomization of upstreams with equal priority, but keep in mind that only lasts for up to the `Refresh` duration. It's still randomized, just not per-request, and this is arguably more efficient too. @jjiang-stripe do you have a preference here? Do you need the randomization of backends within the same priority to be changed at every single request or is at every interval sufficient?

(If it does have to be changed for every request, we need to be mindful that we get upstreams every loop iteration, multiple times per request...)

Reply:
Other users may want to re-query every time, so I suppose it would make sense to allow a `-1` Refresh to turn off caching?

Reply:
Based on a call this morning, sounds like caching this randomization will not be a problem. It could actually be beneficial for RR load balancing. Will leave this as-is unless determined otherwise.

Reply:
I think we need to consider other users too though, especially since this refactor it meant to resolve a bunch of open issues.

Reply:
That's fine, but one thing at a time. Let's see if it really causes a problem for real use cases before we add complexity to a growing system.

Reply:
I'm worried about this being a regression for the few users who are using SRV already. But fair enough.